src/librekey/key_store_g10.c

```diff
@@ -779,14 +1164,218 @@ write_seckey(s_exp_t *s_exp, pgp_seckey_t *key)
+
+    key->encrypted = malloc(key->encrypted_len);
+    if (key->encrypted == NULL) {
+        (void) fprintf(stderr, "can't allocate memory\n");
+        destroy_s_exp(&raw_s_exp);
+        pgp_memory_release(&raw);
+        return false;
+    }
+
+    if (botan_cipher_init(
+          &encrypt, format->botan_cipher_name, BOTAN_CIPHER_INIT_FLAG_ENCRYPT)) {
+        (void) fprintf(stderr, "botan_cipher_init failed\n");
+        goto error;
+    }
+
+    if (botan_cipher_set_key(encrypt, derived_key, keysize)) {
+        goto error;
+    }
+
+    if (botan_cipher_start(encrypt, key->iv, format->iv_size)) {
+        goto error;
+    }
+
+    if (botan_cipher_update(encrypt,
+                            BOTAN_CIPHER_UPDATE_FLAG_FINAL,
+                            key->encrypted,
+                            key->encrypted_len,
+                            &output_written,
+                            raw.buf,
+                            raw.length,
+                            &input_consumed)) {
+        (void) fprintf(stderr, "botan_cipher_update failed\n");
+        goto error;
+    }
+
+    destroy_s_exp(&raw_s_exp);
+    pgp_memory_release(&raw);
+    botan_cipher_destroy(encrypt);
+
+    key->encrypted_len = output_written;
+
+    time(&now);
+    strftime(
+      (char *) key->protected_at, sizeof(key->protected_at), "%Y%m%dT%H%M%S", gmtime(&now));
+
+write:
+    if (!add_sub_sexp_to_sexp(s_exp, &sub_s_exp)) {
+        return false;
+    }
+
+    if (!add_string_block_to_sexp(sub_s_exp, "protected")) {
+        return false;
+    }
+
+    if (!add_string_block_to_sexp(sub_s_exp, format->g10_type)) {
+        return false;
+    }
+
+    if (!add_sub_sexp_to_sexp(sub_s_exp, &sub_sub_s_exp)) {
+        return false;
+    }
+
+    if (!add_sub_sexp_to_sexp(sub_sub_s_exp, &sub_sub_sub_s_exp)) {
+        return false;
+    }
+
+    if (!add_string_block_to_sexp(sub_sub_sub_s_exp, "sha1")) {
+        return false;
+    }
+
+    if (!add_block_to_sexp(sub_sub_sub_s_exp, key->salt, PGP_SALT_SIZE)) {
+        return false;
+    }
+
+    if (!add_unsigned_block_to_sexp(sub_sub_sub_s_exp, key->s2k_iterations)) {
+        return false;
+    }
+
+    if (key->cipher_mode == PGP_CIPHER_MODE_CBC) {
+        if (!add_block_to_sexp(sub_sub_s_exp, key->iv, G10_CBC_IV_SIZE)) {
+            return false;
+        }
+    } else if (key->cipher_mode == PGP_CIPHER_MODE_OCB) {
+        if (!add_block_to_sexp(sub_sub_s_exp, key->iv, G10_OCB_NONCE_SIZE)) {
+            return false;
+        }
+    }
+
+    if (!add_block_to_sexp(sub_s_exp, key->encrypted, key->encrypted_len)) {
+        return false;
+    }
+
+    if (!add_sub_sexp_to_sexp(s_exp, &sub_s_exp)) {
+        return false;
+    }
+
+    if (!add_string_block_to_sexp(sub_s_exp, "protected-at")) {
+        return false;
+    }
+
+    if (!add_block_to_sexp(sub_s_exp, (uint8_t *) key->protected_at, PGP_PROTECTED_AT_SIZE)) {
+        return false;
+    }
+
+    return true;
```

Comment:
This would leak things though. I would change the `error` label to a `done` label that frees everything, and replace all the `return true`/`return false` with `goto done`. Something like what `pgp_eddsa_sign_hash` and a few other functions do: https://github.com/riboseinc/rnp/blob/c7c3d1973156ff0bc82c8b969a2ed0613f7a2fa5/src/lib/eddsa.c#L174